nucypher/crypto/kits.py

```diff
@@ -17,9 +17,10 @@
 from constant_sorrow.constants import UNKNOWN_SENDER, NOT_SIGNED
 from bytestring_splitter import BytestringKwargifier, VariableLengthBytestring
 from nucypher.crypto.splitters import key_splitter, capsule_splitter
+from nucypher.primitives import VersionedBytes
 
 
-class CryptoKit:
+class CryptoKit(VersionedBytes):
```

Comment:
Phew - I can appreciate the semi-polymorphic action implemented via inheritance here, but I'm unsure I agree that a `CryptoKit` or `TreasureMap` *is a type of* `VersionedBytes` since they also exist in non-serialized form.

Curious: Is it instead possible for versioned serializable entities to *have* a `VersionedBytes` or similar class resolver *without* using inheritance here?

Reply:
Agreed. It feels like you could simply use a `VersionedBytes` helper for the `__bytes__` and `from_bytes` functions since the same class does the serialization and deserialization - I don't think anything cares whether some object/class is an instance of `VersionedBytes` - but I can accept that I may be missing nuanced aspects of this.

Reply:
Yeah you guys are right, I'm giving that a try.

Reply:
@derekpierre @KPrasch I'm having trouble working out how to do make this change because of this one  specific construct.

One major piece of the mechanics here which is a big part of making this work seamlessly is that we use this `__new__` method.

```
def __new__(cls, *args, **kwargs):
        """
        When instantiating a specific version of a versioned Bytestring,
        we want to automatically instantiate the latest version without
        requiring implementors to have to deal with any of this.

        If we are instantiating a specifically versioned class, allow for that also
        """

        if not hasattr(cls, 'version'):
            versioned_subclasses = ByteVersioner.__get_versioned_subclasses(cls)
            # take the highest
            cls = sorted(versioned_subclasses, key=lambda x: x.version)[-1]
        return super(VersionedBytes, cls).__new__(cls)
```

This makes it so when you do something like `messagekit = MessageKit(*args, **kwargs)` you get a MessageKitVersion1 or if you do `MessageKit.from_bytes(<v5>...<bytes>)` you get a MessageKitVersion5.

Can one of you suggest a structure to do this as a helper method without using `__new__` that will have the same benefits?

I mean... this would be ok I guess.
```
messagekit = MessageKit.version_util.latest(*args, **kwargs)
```
and then in `to_bytes`
```
self.version_util.prepend_version(output_bytes)
```

The only real difference would be that we'd have to always update the
`UmbralMessageKit` reference to make sure it's the latest version, but this itself to me is an inferior situation with only arguable stylistic benefits.

What do you guys think about all this?

Reply:
Maybe it's more of a naming problem.
`ByteVersioningHelperMixin`?

Reply:
No doubt this *seems* like a mixin, however, `VersionedBytes` performs an operation the subclass *requires*, which is not in the spirit of a mixin for some added functionality.

I think it's possible to keep the `VersionedBytes` class (perhaps renaming it to `VersionedSerializer`), while decoupling the versioning of *serialization* independently from the entity itself.  

What does this look like? Instead of having multiple `MessageKit` classes, we'd end up with multiple versioned `MessageKitSerializer`s that can be composed *on* a `MessageKit` 

IMO this separation of logic will end up being more testable, accurate and maintainable since it draws a clear separation between serialization vs native-object abstraction layers and only one `MessageKit` class needs to be maintained.  

In fact - you can already see evidence of this pattern emerging in your test wherein `MessageKitVersion19` *only* needs to override serialization methods `__bytes__` and `splitter`.

Reply:
> In fact - you can already see evidence of this pattern emerging in your test wherein MessageKitVersion19 only needs to override serialization methods __bytes__ and splitter.

This is not true.   I specially wrote [this test](https://github.com/nucypher/nucypher/pull/1878/files#diff-77ab6f80bf0c9dcad115c140306fcac2R127) to show that this implementation supports arbitrary addition of methods or any other changes a class might need over the hopefully long future.

I understand what you're saying but I think that it's slightly presumptuous to say "the only thing that will ever change on our classes will be the byte splitter and the __bytes__ method".  I think that to sacrifice the flexibility we get from the way this is setup in the service of some arguable compositional style benefits is not the right decision.

That being said @KPrasch I defer to your view of the scope here and will gladly make this change if you specifically acknowledge these concerns I have laid out here.

Reply:
> the only thing that will ever change on our classes will be the byte splitter and the bytes method"

For the scope of `VersionedBytes`, this is actually True - The implementation and methods of the entity being serialized is independent from the serializer's.

Reply:
> I think that to sacrifice the flexibility we get from the way this is setup...

The design I proposed above does not sacrifice this functionality, rather, builds upon your excellent engineering.

>  ...in the service of some arguable compositional style benefits is not the right decision.

This is more than style - My concerns are a matter of API deigned in a abstraction layer that is both accurate and maintainable.

Reply:
👍 on it.

Reply:
In any case - I think this consideration deserves the attention of @jMyles and @cygnusv :-)